File: test_all_apis.js
#!/usr/bin/env node

const axios = require('axios');
const fs = require('fs');

const BASE_URL = 'http://localhost:3001/api';

// Define all API endpoints to test
const API_ENDPOINTS = {
  'Health': '/health',
  'Auth Status': '/auth/status',
  'Market Overview': '/market/overview',
  'Market Sectors': '/market/sectors',
  'Market Indices': '/market/indices',
  'Market Movers': '/market/movers',
  'Dashboard Summary': '/dashboard/summary',
  'Dashboard Metrics': '/dashboard/metrics',
  'Portfolio Summary': '/portfolio',
  'Portfolio Holdings': '/portfolio/holdings',
  'Portfolio Performance': '/portfolio/performance',
  'Stocks List': '/stocks',
  'Stocks Popular': '/stocks/popular',
  'Stock Detail AAPL': '/stocks/AAPL',
  'Stock Detail MSFT': '/stocks/MSFT',
  'Stock Price AAPL': '/stocks/AAPL/price',
  'Stock Financials AAPL': '/stocks/AAPL/financials',
  'Stock Technical AAPL': '/stocks/AAPL/technical',
  'Screener Default': '/screener',
  'Screener Results': '/screener/results',
  'Analytics Portfolio': '/analytics/portfolio',
  'Analytics Performance': '/analytics/performance',
  'Technical Indicators': '/technical/indicators',
  'Technical Analysis': '/technical/analysis',
  'Sentiment Analysis': '/sentiment/analysis',
  'Sentiment Summary': '/sentiment/summary',
  'News Latest': '/news/latest',
  'News Stock AAPL': '/news/AAPL',
  'Calendar Earnings': '/calendar/earnings',
  'Calendar Events': '/calendar/events',
  'Scores Overview': '/scores',
  'Scores Stock AAPL': '/scores/AAPL',
  'Metrics Market': '/metrics/market',
  'Metrics Performance': '/metrics/performance',
  'Risk Analysis': '/risk/analysis',
  'Risk Portfolio': '/risk/portfolio',
  'Settings User': '/settings',
  'Settings Preferences': '/settings/preferences',
  'Sectors Overview': '/sectors',
  'Sectors Performance': '/sectors/performance',
  'Signals Trading': '/signals/trading',
  'Signals Buy': '/signals/buy',
  'Signals Sell': '/signals/sell',
  'Backtest Results': '/backtest/results',
  'Financial Statements': '/financials/statements',
  'Financial Ratios': '/financials/ratios',
  'Trading Positions': '/trading/positions',
  'Trading History': '/trading/history',
  'Price Data AAPL': '/price/AAPL',
  'Price Historical AAPL': '/price/AAPL/historical',
  'Performance Metrics': '/performance/metrics',
  'Performance Summary': '/performance/summary',
  'Watchlist Default': '/watchlist',
  'Data Sources': '/data/sources',
  'Data Status': '/data/status',
  'Economic Indicators': '/economic/indicators',
  'Economic Data': '/economic/data',
  'Commodities Prices': '/commodities/prices',
  'Commodities Overview': '/commodities',
  'ETF List': '/etf',
  'ETF Popular': '/etf/popular',
  'Insider Trading': '/insider/trading',
  'Insider Activity': '/insider/activity',
  'Dividend Calendar': '/dividend/calendar',
  'Dividend History': '/dividend/history',
  'Live Data Stream': '/live-data/stream',
  'Live Data Quotes': '/live-data/quotes',
  'Orders Active': '/orders/active',
  'Orders History': '/orders/history',
  'Analysts Recommendations': '/analysts/recommendations',
  'Analysts Upgrades': '/analysts/upgrades',
  'Earnings Estimates': '/earnings/estimates',
  'Earnings History': '/earnings/history',
  'Positioning Data': '/positioning/data',
  'Strategy Builder': '/strategyBuilder/strategies',
  'Diagnostics System': '/diagnostics/system',
  'Diagnostics Performance': '/diagnostics/performance'
};

async function testEndpoint(name, endpoint) {
  try {
    const startTime = Date.now();
    const response = await axios.get(`${BASE_URL}${endpoint}`, {
      timeout: 10000,
      headers: {
        'Accept': 'application/json',
        'User-Agent': 'API-Test-Client/1.0'
      }
    });
    const endTime = Date.now();
    const responseTime = endTime - startTime;

    return {
      name,
      endpoint,
      status: 'SUCCESS',
      statusCode: response.status,
      responseTime: `${responseTime}ms`,
      dataSize: JSON.stringify(response.data).length,
      hasData: response.data && (Array.isArray(response.data) ? response.data.length > 0 : Object.keys(response.data).length > 0),
      dataStructure: Array.isArray(response.data) ? 'Array' : typeof response.data,
      preview: JSON.stringify(response.data).substring(0, 200) + '...'
    };
  } catch (error) {
    return {
      name,
      endpoint,
      status: 'ERROR',
      statusCode: error.response?.status || 'N/A',
      responseTime: 'N/A',
      error: error.response?.data?.message || error.message,
      errorType: error.code || 'Unknown'
    };
  }
}

async function runAllTests() {
  console.log('🚀 Starting comprehensive API testing...\n');

  const results = [];
  let successCount = 0;
  let errorCount = 0;

  const endpoints = Object.entries(API_ENDPOINTS);

  for (let i = 0; i < endpoints.length; i++) {
    const [name, endpoint] = endpoints[i];
    console.log(`Testing ${i + 1}/${endpoints.length}: ${name} (${endpoint})`);

    const result = await testEndpoint(name, endpoint);
    results.push(result);

    if (result.status === 'SUCCESS') {
      successCount++;
      console.log(`✅ ${result.statusCode} - ${result.responseTime} - ${result.hasData ? 'Has Data' : 'No Data'}`);
    } else {
      errorCount++;
      console.log(`❌ ${result.statusCode} - ${result.error}`);
    }

    // Small delay to prevent overwhelming the server
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  // Generate summary report
  const summary = {
    totalTests: endpoints.length,
    successful: successCount,
    failed: errorCount,
    successRate: ((successCount / endpoints.length) * 100).toFixed(1) + '%',
    timestamp: new Date().toISOString(),
    results: results
  };

  // Save detailed results to file
  fs.writeFileSync('api_test_results.json', JSON.stringify(summary, null, 2));

  console.log('\n📊 TEST SUMMARY:');
  console.log(`Total Endpoints: ${summary.totalTests}`);
  console.log(`Successful: ${summary.successful}`);
  console.log(`Failed: ${summary.failed}`);
  console.log(`Success Rate: ${summary.successRate}`);

  console.log('\n🔍 DETAILED RESULTS:');
  results.forEach(result => {
    if (result.status === 'SUCCESS') {
      console.log(`✅ ${result.name}: ${result.statusCode} (${result.responseTime}) - ${result.hasData ? 'Data Available' : 'No Data'}`);
    } else {
      console.log(`❌ ${result.name}: ${result.statusCode} - ${result.error}`);
    }
  });

  console.log('\n📄 Full results saved to: api_test_results.json');

  return summary;
}

// Run the tests
runAllTests().catch(console.error);

File: test_all_apis.py
#!/usr/bin/env python3
"""
Comprehensive API Testing Suite
Tests all endpoints and identifies issues
"""
import requests
import json
import time
from datetime import datetime

BASE_URL = "http://localhost:3001"
TIMEOUT = 10

class APITester:
    def __init__(self, base_url):
        self.base_url = base_url
        self.results = []
        self.passed = 0
        self.failed = 0

    def test_endpoint(self, method, endpoint, name, expected_status=200):
        """Test a single endpoint"""
        try:
            url = f"{self.base_url}{endpoint}"
            start = time.time()

            if method == "GET":
                response = requests.get(url, timeout=TIMEOUT)
            else:
                response = requests.post(url, timeout=TIMEOUT)

            elapsed = (time.time() - start) * 1000  # ms

            # Check response
            success = response.status_code == expected_status
            status_ok = "✅" if success else "❌"

            # Try to parse JSON
            try:
                data = response.json()
                data_size = len(json.dumps(data))
            except:
                data = None
                data_size = len(response.text)

            result = {
                'endpoint': endpoint,
                'name': name,
                'method': method,
                'status': response.status_code,
                'expected': expected_status,
                'success': success,
                'elapsed_ms': elapsed,
                'data_bytes': data_size,
                'data': data
            }

            self.results.append(result)

            if success:
                self.passed += 1
                print(f"{status_ok} {method:4s} {endpoint:50s} | {response.status_code} | {elapsed:6.0f}ms | {data_size:8d}B")
            else:
                self.failed += 1
                print(f"{status_ok} {method:4s} {endpoint:50s} | {response.status_code} (expected {expected_status}) | {elapsed:6.0f}ms")
                if data:
                    print(f"   Error: {data}")

            return result

        except requests.exceptions.Timeout:
            self.failed += 1
            print(f"❌ {method:4s} {endpoint:50s} | TIMEOUT after {TIMEOUT}s")
            return None
        except requests.exceptions.ConnectionError:
            self.failed += 1
            print(f"❌ {method:4s} {endpoint:50s} | CONNECTION ERROR")
            return None
        except Exception as e:
            self.failed += 1
            print(f"❌ {method:4s} {endpoint:50s} | ERROR: {str(e)}")
            return None

    def run_all_tests(self):
        """Run all API tests"""
        print("\n" + "="*100)
        print("API ENDPOINT TESTING - " + datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
        print("="*100 + "\n")

        # Health checks
        print("📋 HEALTH CHECKS")
        print("-"*100)
        self.test_endpoint("GET", "/health", "Health Check")
        self.test_endpoint("GET", "/api/health", "API Health Check")
        print()

        # Stock Scores endpoints
        print("📊 STOCK SCORES ENDPOINTS")
        print("-"*100)
        self.test_endpoint("GET", "/api/scores", "All Stock Scores")
        self.test_endpoint("GET", "/api/scores?limit=10", "Stock Scores Limited")
        self.test_endpoint("GET", "/api/scores/AAPL", "Single Stock - AAPL")
        self.test_endpoint("GET", "/api/scores/MSFT", "Single Stock - MSFT")
        self.test_endpoint("GET", "/api/scores/INVALID", "Invalid Stock Symbol", 404)
        print()

        # Sector endpoints
        print("🏭 SECTOR ENDPOINTS")
        print("-"*100)
        self.test_endpoint("GET", "/api/sectors", "All Sectors")
        self.test_endpoint("GET", "/api/sectors/Technology", "Single Sector - Technology")
        self.test_endpoint("GET", "/api/sectors/Healthcare", "Single Sector - Healthcare")
        print()

        # Dashboard endpoints
        print("📈 DASHBOARD ENDPOINTS")
        print("-"*100)
        self.test_endpoint("GET", "/api/dashboard", "Dashboard Summary")
        self.test_endpoint("GET", "/api/dashboard/top-movers", "Top Movers")
        print()

        # Analysis endpoints
        print("🔍 ANALYSIS ENDPOINTS")
        print("-"*100)
        self.test_endpoint("GET", "/api/analysis/correlations", "Correlations")
        self.test_endpoint("GET", "/api/analysis/sector-performance", "Sector Performance")
        print()

        # Print summary
        print("\n" + "="*100)
        print(f"TEST RESULTS: {self.passed} passed, {self.failed} failed")
        print("="*100 + "\n")

        return self.results

    def analyze_failures(self):
        """Analyze failed tests"""
        failures = [r for r in self.results if not r.get('success', False)]

        if not failures:
            print("✅ All tests passed!\n")
            return

        print("\n" + "="*100)
        print("FAILED ENDPOINTS - ANALYSIS")
        print("="*100 + "\n")

        for failure in failures:
            print(f"❌ {failure['method']} {failure['endpoint']}")
            print(f"   Expected: {failure['expected']}, Got: {failure['status']}")
            if failure['data']:
                print(f"   Response: {json.dumps(failure['data'], indent=2)[:200]}")
            print()

if __name__ == "__main__":
    # Check if backend is running
    try:
        requests.get(f"{BASE_URL}/health", timeout=2)
    except:
        print(f"❌ Backend not responding at {BASE_URL}")
        print("Please start the backend with: npm start")
        exit(1)

    # Run tests
    tester = APITester(BASE_URL)
    tester.run_all_tests()
    tester.analyze_failures()
